### PR TITLE
Query Parameter Services

### LINKED ISSUE #562: Rework QueryParameters
from #558

### PR DESCRIPTION
**Pull request by maurei**
Closes #562

This PR contains a rework of the internal handling of query parameters. Apart from refactoring, cleaning up and adding unit tests, it contains the following major changes:
- Decoupled query parameter services
- Migrated responsibilities
- Easier extensibility

These are now discussed below

#### Decoupled query parameter services
Every supported url query parameter now has an associated dedicated service for parsing it and internally exposing it.

| Query Parameter Service | Occurence in URL               | Domain                                                |
|-------------------------|--------------------------------|-------------------------------------------------------|
| `IFilterService`         | `?filter[article.title]=title`    | filtering the resultset                               |
| `IIncludeService`         | `?include=article.author`        | including related data                                |
| `IPageService`            | `?page[size]=10&page[number]=3`  | pagination of the resultset                           |
| `ISortService`            | `?sort=-title`                   | sorting the resultset                                 |
| `ISparseFieldsService`    | `?fields[article]=title,summary` | sparse field selection                                |
| `IOmitDefaultService`     | `?omitDefault=true`              | omitting default values from the serialization result, eg `guid-value": "00000000-0000-0000-0000-000000000000"` |
| `IOmitNullService`        | `?omitNull=false`                | omitting null values from the serialization result    |

Throughout the framework, each of these services are now injected only in the places where they are needed. For example: the serialization layer does no longer have access to the filter, sort, page queries.

#### Migrated responsibilities
As mentioned in the previous part, these dedicated services are responsible for parsing and internally exposing the query parameter. This implies that the following migrations of responsibilities have been introduced

- the previously existing `IQueryParser` service is no longer responsible for parsing the url query parameters. The parsing logic for each query has been moved to it's dedicated service (through the implementation of the  `IQueryParameter` interface).
    - eg only the `IFilterService` should know how to parse a `filter[X]=Y` query string
- `IQueryParser` is now `IQueryParameterDiscovery` which is only responsible for discovering the registered query parameter services and providing them with the appropriate part of the URL query string 
    - eg it calles the `IFilterServices.Parse` method with the `filter[X]=Y` part of the query string.
- the repository layer used to be responsible for validating parsed values of query parameters. This responsibility too has been moved to the dedicated query parameter service.
    - eg the `Include` method used to be called with a list of strings (like "article.author.invalid-relationship"), and this list would then be validated and converted into the `RelationshipsAttribute` that is actually reflected by that relationship chian. This type of validation now happens directly upon parsing the query parameters in the middleware

#### Better extensibility
To register a custom query parameter, all one needs to do is register an implementation of `IQueryParameterService`. See the [Query Parameter Services wiki](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/feat/queryparams/wiki/v4/content/query-parameter-services.md).

### CODE REVIEW

File: src/JsonApiDotNetCore/Data/IEntityWriteRepository.cs [hunk 1]
```diff
@@ -17,9 +17,6 @@ public interface IEntityWriteRepository<TEntity, in TId>
 
         Task<TEntity> UpdateAsync(TEntity entity);
 
-        [Obsolete("Use overload UpdateAsync(TEntity updatedEntity): providing parameter ID does no longer add anything relevant")]
```
Comment (maurei) [inline src/JsonApiDotNetCore/Data/IEntityWriteRepository.cs]: Might as well delete deprecated members in upcoming release

File: src/JsonApiDotNetCore/Data/DefaultEntityRepository.cs [hunk 2]
```diff
@@ -224,12 +209,6 @@ public void DetachRelationshipPointers(TEntity entity)
             }
         }
 
-        [Obsolete("Use overload UpdateAsync(TEntity updatedEntity): providing parameter ID does no longer add anything relevant")]
-        public virtual async Task<TEntity> UpdateAsync(TId id, TEntity updatedEntity)
-        {
```
Comment (maurei) [inline src/JsonApiDotNetCore/Data/DefaultEntityRepository.cs]: Might as well delete deprecated members in upcoming release

File: .editorconfig [hunk 3]
```diff
@@ -12,3 +12,14 @@ charset = utf-8
 
 [*.{csproj,props}]
 indent_size = 2
+
+[*.{cs,vb}]
```
Comment (maurei) [inline .editorconfig]: Ensures that auto-initialization of private members will be of the underscore format by default.

File: src/JsonApiDotNetCore/Extensions/IQueryableExtensions.cs [hunk 4]
```diff
@@ -30,46 +30,57 @@ private static MethodInfo ContainsMethod
             }
         }
 
-        public static IQueryable<TSource> Sort<TSource>(this IQueryable<TSource> source, ContextEntity primaryResource, IContextEntityProvider provider, List<SortQuery> sortQueries)
+        public static IQueryable<T> PageForward<T>(this IQueryable<T> source, int pageSize, int pageNumber)
```
Comment (maurei) [inline src/JsonApiDotNetCore/Extensions/IQueryableExtensions.cs]: This part of the diff does not make sense. I did not rename the sort function to the page forward function

File: src/JsonApiDotNetCore/Extensions/IQueryableExtensions.cs [hunk 5]
```diff
@@ -113,52 +124,39 @@ private static IOrderedQueryable<TSource> CallGenericOrderMethod<TSource>(IQuery
...
-        }
-
-        private static Expression GetFilterExpressionLambda(Expression left, Expression right, FilterOperations operation)
+        private static Expression GetFilterExpressionLambda(Expression left, Expression right, FilterOperation  operation)
         {
             Expression body;
             switch (operation)
             {
-                case FilterOperations.eq:
+                case FilterOperation.eq:
```
Comment (maurei) [inline src/JsonApiDotNetCore/Extensions/IQueryableExtensions.cs]: Consistency and best practise: use singular words for enums

File: src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -197,12 +195,12 @@ private static void AddMvcOptions(MvcOptions options, JsonApiOptions config)
             services.AddScoped<IJsonApiReader, JsonApiReader>();
             services.AddScoped<IGenericProcessorFactory, GenericProcessorFactory>();
             services.AddScoped(typeof(GenericProcessor<>));
-            services.AddScoped<IQueryAccessor, QueryAccessor>();
```
Comment (maurei) [inline src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs]: These are fully deprecated in their old forms

File: src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -215,9 +213,20 @@ private static void AddMvcOptions(MvcOptions options, JsonApiOptions config)
...
             services.AddScoped<ISparseFieldsService, SparseFieldsService>();
-            services.AddScoped<IPageQueryService, PageService>();
-
+            services.AddScoped<IPageService, PageService>();
+            services.AddScoped<IOmitDefaultService, OmitDefaultService>();
+            services.AddScoped<IOmitNullService, OmitNullService>();
+
+            services.AddScoped<IQueryParameterService>(sp => sp.GetService<IIncludeService>());
+            services.AddScoped<IQueryParameterService>(sp => sp.GetService<IFilterService>());
+            services.AddScoped<IQueryParameterService>(sp => sp.GetService<ISortService>());
```
Comment (maurei) [inline src/JsonApiDotNetCore/Extensions/IServiceCollectionExtensions.cs]: This is where the "natively supported" query parameter services are registered internally for discovery

File: src/JsonApiDotNetCore/Internal/Query/BaseQueryContext.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
+using JsonApiDotNetCore.Models;
+
+namespace JsonApiDotNetCore.Internal.Query
+{
+    /// <summary>
+    /// A context class that provides extra meta data for a <see cref="TQuery"/>
+    /// that is used when applying url queries internally.
+    /// </summary>
```
Comment (maurei) [inline src/JsonApiDotNetCore/Internal/Query/BaseQueryContext.cs]: This is not an entirely new class, merely a refactor of the previously `RelatedAttrFilterQuery` and `AttributeAttrFilterQuery`

File: src/JsonApiDotNetCore/Middleware/QueryParameterFilter.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
+using System.Reflection;
+using System.Threading.Tasks;
+using JsonApiDotNetCore.Controllers;
+using JsonApiDotNetCore.Services;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace JsonApiDotNetCore.Middleware
+{
+    public class QueryParameterActionFilter : IAsyncActionFilter, IQueryParameterActionFilter
```
Comment (maurei) [inline src/JsonApiDotNetCore/Middleware/QueryParameterFilter.cs]: Part of the middleware that parses query parameters.

File: src/JsonApiDotNetCore/QueryParameterServices/Common/QueryParameterParser.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+using JsonApiDotNetCore.Configuration;
+using JsonApiDotNetCore.Controllers;
+using JsonApiDotNetCore.Internal;
+using JsonApiDotNetCore.Query;
+using Microsoft.AspNetCore.Http;
+
+namespace JsonApiDotNetCore.Services
+{
+    /// <inheritdoc/>
+    public class QueryParameterDiscovery : IQueryParameterDiscovery
```
Comment (maurei) [inline src/JsonApiDotNetCore/QueryParameterServices/Common/QueryParameterParser.cs]: This is the discovery service that pulls in all the registered query parameter services (see constructor) and calls the parse method on them.

File: src/JsonApiDotNetCore/RequestServices/CurrentRequest.cs [hunk 11]
```diff
@@ -1,29 +1,15 @@
...
-using System.Collections.Generic;
 
 namespace JsonApiDotNetCore.Managers
 {
-
     class CurrentRequest : ICurrentRequest
     {
         private ContextEntity _contextEntity;
         public string BasePath { get; set; }
-        public List<string> IncludedRelationships { get; set; }
```
Comment (maurei) [inline src/JsonApiDotNetCore/RequestServices/CurrentRequest.cs]: Removed things from former request manager that are no longer referenced

File: src/JsonApiDotNetCore/Serialization/Server/Builders/LinkBuilder.cs [hunk 12]
```diff
@@ -83,9 +80,7 @@ private string GetSelfTopLevelLink(string resourceName)
 
         private string GetPageLink(ContextEntity primaryResource, int pageOffset, int pageSize)
         {
-            var filterQueryComposer = new QueryComposer();
-            var filters = filterQueryComposer.Compose(_currentRequest);
```
Comment (maurei) [inline src/JsonApiDotNetCore/Serialization/Server/Builders/LinkBuilder.cs]: Removed this as it is incomplete and untested.

**Comment by wisepotato** [comment]
We should consider adding `Query` in each of the service names.

**Comment by wisepotato** [comment]
wiki is clear however it looks like i can only insert one extra paramservice, is this true? how do i add two? would i have: 

```cs
services.AddScoped<IQueryParameterService, TeapotService>();
services.AddScoped<IQueryParameterService, OtherTeaPotService>();
```
?

**Pull request merged**